app/services/patient_import/importer.rb

```diff
@@ -0,0 +1,80 @@
+class PatientImport::Importer
+  include Api::V3::SyncEncounterObservation
+  include Api::V3::RetroactiveDataEntry
+
+  attr_reader :params, :facility, :admin, :results
+
+  def self.call(*args)
+    new(*args).call
+  end
+
+  def initialize(params:, facility:, admin:)
+    @params = params
+    @facility = facility
+    @admin = admin
+    @results = []
+  end
+
+  def call
+    ActiveRecord::Base.transaction do
+      params.each_with_index do |patient_record_params, index|
+        patient_params = Api::V3::PatientTransformer.from_nested_request(patient_record_params[:patient])
+        medical_history_params = Api::V3::MedicalHistoryTransformer.from_request(patient_record_params[:medical_history])
+        blood_pressures_params = patient_record_params[:blood_pressures].map { |bp| Api::V3::BloodPressureTransformer.from_request(bp) }
+        prescription_drugs_params = patient_record_params[:prescription_drugs].map { |drug| Api::V3::PrescriptionDrugTransformer.from_request(drug) }
+
+        import_patient(patient_params)
+        import_blood_pressures(blood_pressures_params)
+        import_medical_history(medical_history_params)
+        import_prescription_drugs(prescription_drugs_params)
+
+        results.push(
+          patient: patient_params,
+          medical_history: medical_history_params,
+          blood_pressures: blood_pressures_params,
+          prescription_drugs: prescription_drugs_params
+        )
+      end
```

Comment:
Same question as @ssrihari - what are the expected sizes for import spreadsheets? We probably need to batch up rows if is anything over a few hundred, given the amount of work going on here.  One malformed cell will also cause the entire transaction blow up, so error handling will be tricky. 😟

Reply:
Actually, will failed results from all the `merge` calls result in a rollback at all? If we end up just doing a `save` or `update` down the stack, I think things will silently fail.

Reply:
There are 100-200 patients per spreadsheet.

The two options to commit to are:
1. All or nothing
1. Import what you can, report the failures well

The validations that we are doing before import should in theory take care of all problems with the data itself. How, then, can the import fail? Things like timeouts, network blips to DB, and of course edge cases we haven't thought about. I've just confirmed that Rob's suspicion about `merge` failing silently is true. In that case, I'm starting to lean towards approach 2.

Only power users (us) are going to be using this, so I'd like to prep each spreadsheet, upload it to ET Demo, verify if any records failed and why, and then upload to prod.

Additionally, we should be mindful not to open up this dashboard page to other users. We're making a lot of technical decisions here that are hinged upon the assumption that this tool is for our eyes only.

Reply:
I've leaned into this idea with my latest commits.
* Every patient gets a separate job queued
* Every patient gets its operations wrapped in a DB transaction inside the job - This way we never import a patients BPs, but not their drugs, for example
* If a patient import fails because of transient issues, the job will retry and succeed
* If a patient import fails because the data is bad, and somehow got through our validations, then the job will stick around in Sidekiq, and the dev who ran the import can inspect the failed jobs there.

I think this is the simplest approach assuming that this is a _developer-only_ tool. What do you think?

Reply:
This sounds like a great approach to me, and will be much easier to troubleshoot when things come up.

Reply:
Agreed, this looks good! Some additional notes/observation:

>  merge failing silently

This seems like something we need to address regardless of which approach we take, ya?

> upload it to ET Demo

Um, would we upload real PHI to another environment? Maybe we can do it locally instead?

> We're making a lot of technical decisions here that are hinged upon the assumption that this tool is for our eyes only.

Let's call this out in a comment next to the permissions code?

Reply:
Actually, if only power users should do this, should we just put it in a rake task? That way it's obvious we shouldn't (can't) allow others to use it.

Reply:
Ah sorry I just remembered you mentioned convenience in this case is important, nvm re rake task. ✌️

Reply:
Exactly re: rake task 😄 

Regarding PHI in a Demo environment, you're right. Demo might not be the right place, but a production-like environment will be important. I've been using the Heroku review app until now, which is great as it will die soon. Regarding sensitivity, we've definitely leaned towards "more secure, less aggressive testing" int he past and it bit us. I think the value of testing the actual PHI in, say, a review app is worth the extra exposure.

Reply:
Re silent merges, yes. I think this is a hot topic for "sync api TLC". I'll add a note to https://app.clubhouse.io/simpledotorg/epic/3366/revisit-sync?ct_workflow=all